keps/sig-node/127-user-namespaces/README.md

```diff
@@ -0,0 +1,456 @@
+  as different UIDs in the node and therefore are more isolated than today.
+  See phase 3 for limitations.
+- Allow pods to have capabilities (e.g. `CAP_SYS_ADMIN`) that are only valid in
+  the pod (not valid in the host).
+- Benefit from the security hardening that user namespaces provide against some
+  of the future unknown runtime and kernel vulnerabilities.
+
+### Non-Goals
+
+- Provide a way to run the kubelet process or container runtimes as an
+  unprivileged process. Although initiatives like [kubelet in user
+  namespaces][kubelet-userns] and this KEP both make use of user namespaces, it is
+  a different implementation for a different purpose.
+- Implement all the very nice use cases that user namespaces allows. The goal
+  here is to allow them as incremental improvements, not implement all the
+  possible ideas related with user namespaces.
+
+[kubelet-userns]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2033-kubelet-in-userns-aka-rootless
+
+## Proposal
+
+This proposal aims to support running pods inside user namespaces. This will
+improve the pod to node isolation (phase 1 and 2) and pod to pod isolation
+(phase 3) we currently have.
+
+This mitigates all the vulnerabilities listed in the motivation section.
+
+### User Stories
+
+#### Story 1
+
+As a cluster admin, I want run some pods with privileged capabilities because
+the applications in the pods require it (e.g. `CAP_SYS_ADMIN` to mount a FUSE
+filesystem or `CAP_NET_ADMIN` to setup a VPN) but I don't want this extra
+capability to grant me any extra privileges on the host or other pods.
+
+#### Story 2
+
+As a cluster admin, I want to allow some pods to run in the host user namespace
+if they need a feature only available in such user namespace, such as loading a
+kernel module with `CAP_SYS_MODULE`.
+
+#### Story 3
+
+As a cluster admin, I want to allow users to run their container as root
+without that process having root privileged on the host, so I can mitigate the
+impact of a compromised container.
+
+#### Story 4
+
+As a cluster admin, I want to allow users to choose the user/group they use
+inside the container without it being the user/group they use on the node, so I
+can mitigate the impact of a container breakout vulnerability (e.g to read
+host files).
+
+#### Story 5
+
+As a cluster admin, I want to use different host UIDs/GIDs for pods running in
+the same node (whenever kernel/kube features allow it), so I can mitigate the
+impact a compromised pod can have on other pods and the node itself.
+
+### Notes/Constraints/Caveats
+
+#### UID space
+
+#### Volume support
+
+### Risks and Mitigations
+
+## Design Details
+
+Note: Names are preliminary yet, I'm using field names to simplify explanations.
+
+### Pod.spec changes
+
+The following changes will be done to the pod.spec:
+
+- `pod.spec.useHostUsers`: bool.
+If true or not present, uses the host user namespace (as today)
```

Comment:
We need to balance the desire to not "opt-in to security" with compatibility.  Ideally we would make the default be "use userns", but that can break a lot of pods.  But if we make this opt-in, uptake will be low.  I think this is worth some amount of risk, but how much?  We can tell admins to set an admission hook to change not-specified into `false`, but we're just shifting the burden.  So how can we do better?

When do KNOW it is safe to enable?

When:
* not using any of {hostNetwork, hostPID, hostIPC}
* not using hostPath volumes
* all volumes are "safe" (as listed elsewhere)

In that case we should be able to default this to `false` - right?.

If any of those are in play (maybe not hostNetwork?) then we we have to default to `hostUsers: true`.

Reply:
I think making it on by default is a different discussion. We can make it enable by default in some limited scenarios. I can add to the KEP that discussion on when to enable it by default is pending, so we don't forget.

However, userns is not a security feature that you can enable and its transparent to the user. You run as a different effective user, so tons of things changes, you don't have permissions to do some stuff (like load kernel modules), files you create (even on safe volumes like an emptyDir, etc.) will have a different UID than the "RunAsUser", etc. We should be _very_ cautious about enabling it by default and is a very important conversation to have.

However, I don't think it is useful to have it know. My idea with this PR is to agree on the high level idea of the feature and move towards alpha, we can later have this discussion when moving to beta and with all the feedback we have from users.

@thockin what do you think?

Reply:
I agree that it should not be enabled by default YET.  I also think that it may be desirable to make the default behaviour a cluster or node tunable.  But this can and should be decided later.  Opt-in on a per-pod basis is good for the first step.

I also think that even though it is opt-in, there are many workloads that will use this feature not specifically for the security benefits, but beacuse you can safely and without special privileges run your workload as "root" and other particular UIDs.  This is important for legacy programs and systems that cannot easily be changed to run under arbitrary UIDs.

Reply:
ACK - let's call it out here, e.g. as a GA graduation requirement  "make plans on whether, when, and how to enable by default"

Reply:
Added that to the GA graduation, will be in the next push!